Update minimum java version to 21

4drian3d opened the pull request:
resolves #1601 
closes #1462 
resolves #1320

4drian3d committed:
Update minimum java version to 21

Timongcraft commented:
Do you think it would be appropriate to also migrate to pattern matching for `instanceof` in this PR?

4drian3d committed:
Use modern java features in more places

4drian3d committed:
Use modern java features in more places

4drian3d committed:
Use modern java features in more places

4drian3d committed:
Removed log4j2 plugin module

The log4j2 plugin transformer bug in shadow was resolved in version 9.x, so this module is no longer necessary

4drian3d committed:
Updated some dependencies

4drian3d committed:
Merge branch 'dev/3.0.0' into update/java/21

4drian3d committed:
Updated gson and guava to align with the versions used by Minecraft Vanilla

resolves #1601

4drian3d committed:
Updated log4j2 version

closes #1462

Emilxyz commented on proxy/src/main/java/com/velocitypowered/proxy/protocol/packet/LegacyPlayerListItemPacket.java:
```diff
@@ -69,33 +69,25 @@ public void decode(ByteBuf buf, ProtocolUtils.Direction direction, ProtocolVersi
-              if (buf.readBoolean()) {
-                item.setPlayerKey(ProtocolUtils.readPlayerKey(version, buf));
-              }
+                if (buf.readBoolean()) {
+                    item.setPlayerKey(ProtocolUtils.readPlayerKey(version, buf));
+                }
```
this got changed to 4 spaces indentation, or are my eyes playing a joke on me?

Emilxyz commented on proxy/src/main/java/com/velocitypowered/proxy/protocol/packet/LegacyPlayerListItemPacket.java:
```diff
@@ -69,33 +69,25 @@ public void decode(ByteBuf buf, ProtocolUtils.Direction direction, ProtocolVersi
+          case UPDATE_LATENCY -> item.setLatency(ProtocolUtils.readVarInt(buf));
+          case UPDATE_DISPLAY_NAME -> item.setDisplayName(readOptionalComponent(buf, version));
+          case REMOVE_PLAYER -> {
+              //Do nothing, all that is needed is the uuid
```
only a comment, but technically here also

Emilxyz commented on proxy/src/main/java/com/velocitypowered/proxy/protocol/packet/LegacyPlayerListItemPacket.java:
```diff
@@ -126,39 +118,32 @@ public void encode(ByteBuf buf, ProtocolUtils.Direction direction, ProtocolVersi
-              } else {
-                buf.writeBoolean(false);
-              }
+                if (item.getPlayerKey() != null) {
+                    buf.writeBoolean(true);
+                    ProtocolUtils.writePlayerKey(buf, item.getPlayerKey());
+                } else {
+                    buf.writeBoolean(false);
+                }
```
and here

Emilxyz commented on proxy/src/main/java/com/velocitypowered/proxy/protocol/packet/LegacyPlayerListItemPacket.java:
```diff
@@ -126,39 +118,32 @@ public void encode(ByteBuf buf, ProtocolUtils.Direction direction, ProtocolVersi
+          case UPDATE_LATENCY -> ProtocolUtils.writeVarInt(buf, item.getLatency());
+          case UPDATE_DISPLAY_NAME -> writeDisplayName(buf, item.getDisplayName(), version);
+          case REMOVE_PLAYER -> {
+              // Do nothing, all that is needed is the uuid
```
only a comment again, but also here

Emilxyz commented on proxy/src/main/java/com/velocitypowered/proxy/protocol/packet/chat/ComponentHolder.java:
```diff
@@ -113,23 +113,32 @@ public BinaryTag getBinaryTag() {
   public static BinaryTag serialize(JsonElement json) {
     if (json instanceof JsonPrimitive jsonPrimitive) {
 		if (jsonPrimitive.isNumber()) {
```
okay it wasn't changed by this pr, but this is using tabs instead of spaces

Emilxyz commented on proxy/src/main/java/com/velocitypowered/proxy/protocol/packet/chat/ComponentHolder.java:
```diff
@@ -113,23 +113,32 @@ public BinaryTag getBinaryTag() {
+                    return IntBinaryTag.intBinaryTag(l.intValue());
+                }
+                default -> {}
+            }
```
indents are weird here as well, I also wonder if it would be a little "cleaner" to just do
```java
return switch (number) {
  case Byte b -> ByteBinaryTag.byteBinaryTag(b);
  case Short s -> // ...
  // ...
  default -> EndBinaryTag.endBinaryTag(); // or maybe throw an exception?
}
```

Emilxyz commented on proxy/src/main/java/com/velocitypowered/proxy/protocol/packet/title/LegacyTitlePacket.java:
```diff
@@ -40,24 +40,19 @@ && getAction() == ActionType.SET_ACTION_BAR) {
+          if (component == null) {
+              throw new IllegalStateException("No component found for " + getAction());
+          }
+          component.write(buf);
```
indents here as well

4drian3d committed:
Apply review comments

4drian3d committed:
Merge branch 'dev/3.0.0' into update/java/21

4drian3d committed:
Expanded external plugin warning to recognize Paper plugins

4drian3d committed:
Merge branch 'dev/3.0.0' into update/java/21

4drian3d committed:
Merge branch 'dev/3.0.0' into update/java/21

# Conflicts:
#	gradle/libs.versions.toml
#	proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java

4drian3d committed:
Fixed sending ServerData packets if the description component from the backend server is null (#1673)

p0t4t0sandwich committed:
Moved pre-1.19.1 command argument validation so it prints the faulty identifier (#1675)

4drian3d committed:
Generate a new forwarding secret file if the file is deleted (#1671)

* Generate a new forwarding secret file if the file is deleted

This allows to generate a new forwarding secret simply by deleting the file if required.
The file will only be generated if the forwarding secret is not configured through a system property

resolves #1670

* Add file creation message

SNWCreations committed:
feat: PlayerChannelUnregisterEvent (#1686)

* feat: PlayerChannelUnregisterEvent

* style: fix checkstyle issues

szymon-off committed:
Fix exception message formatting in command invocation (#1687)

4drian3d committed:
[ci skip] Replaced slf4j javadocs provider with javadocs.dev

4drian3d committed:
Minecraft 1.21.11 (#1690)

* 1.21.11-pre2

I have not identified any changes that affect us at this time

* 1.21.11-pre3

* 1.21.11-pre4

* 1.21.11-pre5

* 1.21.11-rc1

* 1.21.11-rc2

* Minecraft 1.21.11 release support

RealBauHD committed:
bump adventure to 4.26.1 (#1697)

electronicboy committed:
Restrict empty packet frames from clients

electronicboy committed:
Conformity

Warriorrrr committed:
Provide an IDE pattern hint for plugin IDs (#1712)

* Provide an IDE pattern hint for plugin IDs

* order my imports the right way

mxson committed:
Fix ByteBuf memory leak in MinecraftVarintFrameDecoder (#1715)

- Reset buffer reader index on exception to prevent memory leaks when packet decoding fails.

4drian3d committed:
Merge branch 'dev/3.0.0' into update/java/21

# Conflicts:
#	gradle/libs.versions.toml
#	proxy/src/main/java/com/velocitypowered/proxy/connection/util/ServerListPingHandler.java
#	proxy/src/main/java/com/velocitypowered/proxy/protocol/packet/chat/ComponentHolder.java

kennytv committed:
Merge branch 'dev/3.0.0' into update/java/21

kennytv committed:
oooo
